core/client/fs/src/main/java/alluxio/client/file/FileSystemContextReinitializer.java

```diff
@@ -86,49 +76,95 @@ public void onSuccess() {
     mExecutor.resetMetaMasterConfigClient(mContext.getMasterClientContext());
   }
 
+  /**
+   * This resource blocks reinitialization until close.
+   */
+  public static final class BlockerResource implements Closeable {
+    private CountLatch mLatch;
+
+    /**
+     * Increases the count of the latch.
+     * If the latch is held by {@link AllowerResource}, the constructor blocks until the latch is
+     * released by {@link AllowerResource#close()}.
+     *
+     * @param latch the count latch
+     * @throws InterruptedException if interrupted during being blocked
+     */
+    public BlockerResource(CountLatch latch) throws InterruptedException {
+      mLatch = latch;
+      mLatch.inc();
+    }
+
+    @Override
+    public void close() {
+      mLatch.dec();
+    }
+  }
+
+  /**
+   * This resource allows reinitialization.
+   * It blocks further RPCs until close.
+   */
+  public static final class AllowerResource implements Closeable {
+    private CountLatch mLatch;
+
+    /**
+     * Waits the count to reach zero, then blocks further constructions of {@link BlockerResource}
+     * until {@link #close()}.
+     *
+     * @param latch the count latch
+     */
+    public AllowerResource(CountLatch latch) {
+      mLatch = latch;
+      mLatch.awaitAndBlockInc();
+    }
+
+    @Override
+    public void close() {
+      mLatch.unblockInc();
+    }
+  }
+
   /**
    * Acquires the resource to block reinitialization.
    *
-   * When the context is being reinitialized, this call returns an empty optional.
+   * When the context is being reinitialized, this call blocks.
    * If the reinitialization fails, an exception is thrown.
    *
    * If there is an existing reinitialization exception, immediately throw an exception
    * without trying to block further reinitialization.
    *
    * @throws IOException when reinitialization fails before or during this method
+   * @throws InterruptedException if interrupted during being blocked
    * @return the resource
    */
-  public Optional<CountResource> acquireResourceToBlockReinit() throws IOException {
+  public BlockerResource acquireResourceToBlockReinit() throws IOException, InterruptedException {
     Optional<IOException> exception = mExecutor.getException();
     if (exception.isPresent()) {
       throw exception.get();
     }
-    if (mCount.updateAndGet(c -> c >= 0 ? c + 1 : c) <= 0) {
-      return Optional.empty();
-    }
+    BlockerResource r = new BlockerResource(mLatch);
```

Comment:
`InitBlockerResource` and `InitAllowerResource` maybe?
That would complement changing names in `FileSystemContextReinitializer` to `acquireShared/Exclusive`

Reply:
changed to `ReinitBlockerResource` and `ReinitAllowerResource`